DataFormats/CTPPSReco/src/CTPPSPixelLocalTrack.cc

```diff
@@ -0,0 +1,99 @@
+/*
+ *
+* This is a part of CTPPS offline software.
+* Author:
+*   Fabrizio Ferro (ferro@ge.infn.it)
+*   Enrico Robutti (robutti@ge.infn.it)
+*   Fabio Ravera   (fabio.ravera@cern.ch)
+*
+*/
+
+#include "DataFormats/CTPPSReco/interface/CTPPSPixelLocalTrack.h"
+// #include "Rtypes.h"
+
+//----------------------------------------------------------------------------------------------------
+
+math::Matrix<2,2>::type CTPPSPixelLocalTrack::trackPointInterpolationCovariance(float z) const
+{
+  math::Matrix<2, dimension>::type h;
+  h(0,0)=1;
+  h(1,1)=1;
+  h(0,2)=z-z0_;
+  h(1,3)=z-z0_;
+  
+  CovarianceMatrix cov_matr;
+  for(unsigned int i=0; i<dimension; ++i)
+    for(unsigned int j=0; j<dimension; ++j)
+      cov_matr[i][j]=covarianceMatrixElement(i,j);
+
+  math::Matrix<dimension,2>::type  hT = ROOT::Math::Transpose(h);
+  
+  math::Matrix<dimension,2>::type V_hT = cov_matr * hT;//(cov_matr, TMatrixD::kMultTranspose, h);
+  //h*=V_hT;
+  //return h;
+  math::Matrix<2,2>::type h_V_hT = h * V_hT;
+  return h_V_hT;
+}
+
+//----------------------------------------------------------------------------------------------------
+
+CTPPSPixelLocalTrack::CTPPSPixelLocalTrack(float z0, const ParameterVector & track_params_vector, 
+      const CovarianceMatrix &par_covariance_matrix, float chiSquared) 
+      : z0_(z0), chiSquared_(chiSquared), valid_(true), numberOfPointUsedForFit_(0)
+{
+  for(unsigned int i=0; i<dimension; ++i)
+  {
+    track_params_vector_[i]=track_params_vector[i];
+    for(unsigned int j=0; j<dimension; ++j)
+    {
+      covarianceMatrixElement(i,j)=par_covariance_matrix(i,j);
+    }
+  }
+}
+
+//----------------------------------------------------------------------------------------------------
+
+CTPPSPixelLocalTrack::ParameterVector CTPPSPixelLocalTrack::getParameterVector() const 
+{
+  ParameterVector v;
+  
+  for (unsigned int i = 0; i < dimension; ++i)
+    v[i] = track_params_vector_[i];
+      
+  return v;
+}
+
+//----------------------------------------------------------------------------------------------------
+
+CTPPSPixelLocalTrack::CovarianceMatrix CTPPSPixelLocalTrack::getCovarianceMatrix() const 
+{
+  CovarianceMatrix m;
+  
+  for(int i=0; i<dimension; ++i)
+    for(int j=0; j<dimension; ++j)
+      m(i,j) = covarianceMatrixElement(i,j);
+      
+  return m;
+}
+
+//----------------------------------------------------------------------------------------------------
+
+bool CTPPSPixelLocalTrack::operator< (const CTPPSPixelLocalTrack &r)
```

Comment:
remind me, please, why there should be a comparison operator for tracks.

Reply:
Dear Slava,
our understanding was that the operator < is defined.
Moreover we are storing the tracks in a DetSet which should have the sort method.
If you think that we can safely remove it I would proceed.

Reply:
IIUC, the DetSet requires ordering only for the key (the DetId), but not for the content (the CTPPSPixelLocalTracks).

Reply:
Dear Slava, I tried to comment this operator out but I got problem in compiling the CTPPSPixelLocalTrack class:

In file included from input_line_8:49:
In file included from /cvmfs/cms-ib.cern.ch/week1/slc6_amd64_gcc630/cms/cmssw/CMSSW_9_3_X_2017-09-15-2300/src/DataFormats/Common/interface/DetSetVector.h:43:
/cvmfs/cms-ib.cern.ch/nweek-02489/slc6_amd64_gcc630/external/boost/1.63.0-mlhled2/include/boost/concept_check.hpp:244:30: error: invalid operands to binary expression ('CTPPSPixelLocalTrack' and 'CTPPSPixelLocalTrack')
      require_boolean_expr(a < b);
                           ~ ^ ~
/cvmfs/cms-ib.cern.ch/nweek-02489/slc6_amd64_gcc630/external/boost/1.63.0-mlhled2/include/boost/concept/usage.hpp:16:43: note: in instantiation of member function 'boost::LessThanComparable<CTPPSPixelLocalTrack>::~LessThanComparable' requested here
    ~usage_requirements() { ((Model*)0)->~Model(); }
                                          ^
/cvmfs/cms-ib.cern.ch/nweek-02489/slc6_amd64_gcc630/external/boost/1.63.0-mlhled2/include/boost/concept/detail/general.hpp:39:42: note: in instantiation of member function 'boost::concepts::usage_requirements<boost::LessThanComparable<CTPPSPixelLocalTrack> >::~usage_requirements' requested here
    static void failed() { ((Model*)0)->~Model(); }
                                         ^
/cvmfs/cms-ib.cern.ch/nweek-02489/slc6_amd64_gcc630/external/boost/1.63.0-mlhled2/include/boost/concept_check.hpp:243:5: note: in instantiation of member function 'boost::concepts::requirement<boost::concepts::failed ************boost::concepts::usage_requirements<boost::LessThanComparable<CTPPSPixelLocalTrack>
      >::************>::failed' requested here
    BOOST_CONCEPT_USAGE(LessThanComparable) {
    ^
/cvmfs/cms-ib.cern.ch/nweek-02489/slc6_amd64_gcc630/external/boost/1.63.0-mlhled2/include/boost/concept/usage.hpp:29:7: note: expanded from macro 'BOOST_CONCEPT_USAGE'
      BOOST_CONCEPT_ASSERT((boost::concepts::usage_requirements<model>)); \
      ^
/cvmfs/cms-ib.cern.ch/nweek-02489/slc6_amd64_gcc630/external/boost/1.63.0-mlhled2/include/boost/concept/assert.hpp:43:5: note: expanded from macro 'BOOST_CONCEPT_ASSERT'
    BOOST_CONCEPT_ASSERT_FN(void(*)ModelInParens)
    ^
/cvmfs/cms-ib.cern.ch/nweek-02489/slc6_amd64_gcc630/external/boost/1.63.0-mlhled2/include/boost/concept/detail/general.hpp:71:51: note: expanded from macro 'BOOST_CONCEPT_ASSERT_FN'
    &::boost::concepts::requirement_<ModelFnPtr>::failed>    \
                                                  ^
Error: rootcling: compilation failure (tmp/slc6_amd64_gcc630/src/DataFormats/CTPPSReco/src/DataFormatsCTPPSReco/a/DataFormatsCTPPSReco_xr6d2ee5aecf_dictUmbrella.h)
config/SCRAM/GMake/Makefile.rules:2070: recipe for target 'tmp/slc6_amd64_gcc630/src/DataFormats/CTPPSReco/src/DataFormatsCTPPSReco/a/DataFormatsCTPPSReco_xr.cc' failed
gmake: *** [tmp/slc6_amd64_gcc630/src/DataFormats/CTPPSReco/src/DataFormatsCTPPSReco/a/DataFormatsCTPPSReco_xr.cc] Error 1
gmake: *** [There are compilation/build errors. Please see the detail log above.] Error 2

Reply:
now I see in the DetSetVector 
```
    /// DetSetVector requires that T objects can be compared with
    /// operator<.
BOOST_CLASS_REQUIRE(T, boost, LessThanComparableConcept);
```

sorry for the noise.